rfcs/0086-nixos-generation-apis.md

```diff
@@ -0,0 +1,195 @@
+is proposed. Because the old implementation ignored unknown flags and echoed the
+default output, they have to be checked for equality in the list of generations.
+
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## --generation
+Switching to a specific generation:
+```
+# nixos-rebuild switch --generation 121
+```
+The other subcommands like `test`/`boot`/`build` are also working:
+```
+# nixos-rebuild test --generation 122
+```
+
+Other profiles are supported:
+```
+# nixos-rebuild switch --generation 1 --profile-name test
+# nixos-rebuild switch --generation 2 -p test
+```
+
+## list-generations
+```
+$ nixos-rebuild list-generations
+Generation  Build-date               NixOS version               Kernel   Configuration Revision
+121         Sat 2021-02-05 12:25:55  21.03.20210206.9bd6a6c      5.9.9    2199dd2822638f7dce2a9895930c51abd1723234
+122         Sat 2021-02-06 23:19:24  21.03.20210206.e6a8458      5.9.9    dirty
+123         Sat 2021-02-06 23:20:21  21.03.20210206.e6a8458      5.9.9    5e3848ececf21a1ee6b2ac4944bc5db2c02dd281  (current)
+```
+
+Other profiles are supported:
+```
+$ nixos-rebuild list-generations -p test
+Generation  Build-date               NixOS version           Kernel   Configuration Revision
+1           Sat 2021-02-05 12:25:55  21.03.20210206.9bd6a6c  5.9.9    2199dd2822638f7dce2a9895930c51abd1723234
+2           Sat 2021-02-06 23:19:24  21.03.20210206.e6a8458  5.9.9    dirty  (current)
+3           Sat 2021-02-06 23:20:21  21.03.20210206.e6a8458  5.9.9    5e3848ececf21a1ee6b2ac4944bc5db2c02dd281
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## list-generations
+
+- Displaying the `list-generations` in `column` decreases machine-readability (different amount of white-spaces).
+Since the output is primarily intended for humans, this should not be a problem.
+An alternative would be, to follow the `git` model and have `--porcelain` version,
+which is machine readable and guarantees a stable output format.
+- If the configuration revision was a hash of the flake-repository, truncating
+it would make sense. But it could just as well be plain text, where this approach
+would fail. There would be a lot of logic involved for distinguishing between
+hashes and text and conditional truncation.
+- The name `nixos-rebuild` isn't strictly fitting to the `list-generations` parameter
+but a more general `nixos` command as proposed in
+[100578](https://github.com/NixOS/nixpkgs/pull/100578) and
+[54188](https://github.com/NixOS/nixpkgs/issues/54188).
+Though @edolstra [said it would be okay for now](https://github.com/NixOS/nixpkgs/issues/105910#issuecomment-754036275).
+
+
+## nixos-version
+
+Changing `nixos-version` so that a mistyped parameter throws an error improves
+error detection, but is a potentially breaking change, if there were scripts
+relying on mistyped flags. Though my opinion is, that it is better to throw
+meaningful errors instead of just doing something.
```

Comment:
This looks like an independent change. But Is worth it nonetheless.

If scripts mistakenly passed an argument then it's up to them to fix their invocation.

Reply:
Yes, it sort of is independent, but the current implementation will just print the default value for any unrecognised flag and so in the implementation of list-generations I had to [implement a workaraound](https://github.com/NixOS/nixpkgs/blob/4ebca19b171dd5d0d0660ebfd9c413d24a70374e/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L448-L455) that ignores the output if it is the default. I agree -  I probably should have put this in another RFC and kept the default behaviour here...